### PR TITLE
Refactor configurer - Phase 3

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Summary
This PR includes final refactoring configurer procedures.
1. Remove unused parameters from model config to prevent user from confusion
2. Create new configurer mixins for incremental learning and semi supervised learning to reduce duplicated codes across algos.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py [hunk 1]
```diff
@@ -284,49 +222,25 @@ def configure_loss(self, cfg):
             )
             update_or_add_custom_hook(cfg, ib_loss_hook)
 
+    @staticmethod
+    def get_sampler_type(cfg):
+        """Return sampler type."""
+        if not cfg.model.get("multilabel", False) and not cfg.model.get("hierarchical", False):
```
Comment (eunwoosh) [inline src/otx/algorithms/classification/adapters/mmcls/configurer.py]: How about to make this line as a function (such as `_is_multiclass()`?) to increase readability and reusability?

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py [hunk 2]
```diff
@@ -284,49 +222,25 @@ def configure_loss(self, cfg):
...
+            cfg["task_adapt"].get("efficient_mode", True)
+            sampler_type = "balanced"
+        else:
+            cfg["task_adapt"].get("efficient_mode", False)
+            sampler_type = "cls_incr"
+        return sampler_type
```
Comment (goodsong81) [inline src/otx/algorithms/classification/adapters/mmcls/configurer.py]: What are the `get()` in line#229 & 232 doing?

File: src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+class IncrConfigurerMixin:
+    """Patch config to support incremental learning for object detection."""
+
+    org_model_classes: List = []
+    model_classes: List = []
+
+    def configure_task(self, cfg):
+        """Patch config to support incremental learning."""
+        super().configure_task(cfg)
+        if "task_adapt" in cfg and self.task_adapt_type == "mpa":
```
Comment (goodsong81) [inline src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py]: Why don't you take the change to change the name "mpa"? :)

**Review by goodsong81 — DISMISSED** [review]
Thanks for the neat refactoring. A few comments of mine:

File: src/otx/algorithms/common/adapters/mmcv/semisl_mixin.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+class SemiSLConfigurerMixin:
+    """Patch config to support semi supervised learning."""
+
+    def configure_data(self, cfg, data_cfg):
+        """Patch cfg.data."""
+        super().configure_data(cfg, data_cfg)
+        # Set unlabeled data hook
+        if self.training:
+            if cfg.data.get("unlabeled", False) and cfg.data.unlabeled.get("otx_dataset", False):
+                if len(cfg.data.unlabeled.get("pipeline", [])) == 0:
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/semisl_mixin.py:25]:
```suggestion
                if cfg.data.unlabeled.get("pipeline", []):
```
How about to align with PEP recommended way to check a empty list?

File: src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        )
+
+    def is_incremental(self) -> bool:
+        """Return whether current model classes is increased from original model classes."""
+        sampler_flag = True
+        if len(set(self.org_model_classes) & set(self.model_classes)) == 0 or set(self.org_model_classes) == set(
+            self.model_classes
+        ):
+            sampler_flag = False
+        return sampler_flag
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py]:
```suggestion
        return (
            set(self.org_model_classes) & set(self.model_classes)
            and set(self.org_model_classes) != set(self.model_classes)
        )
```
How about to change like this?

File: src/otx/algorithms/segmentation/adapters/mmseg/configurer.py [hunk 6]
```diff
@@ -202,89 +197,27 @@ def configure_input_size(
...
-        task_adapt_hook = ConfigDict(
-            type="TaskAdaptHook",
-            src_classes=self.org_model_classes,
-            dst_classes=self.model_classes,
-            model_type=cfg.model.type,
-            sampler_flag=is_cls_incr,
-            efficient_mode=cfg["task_adapt"].get("efficient_mode", False),
-        )
-        update_or_add_custom_hook(cfg, task_adapt_hook)
+        return is_cls_incr
```
Comment (eunwoosh) [inline src/otx/algorithms/segmentation/adapters/mmseg/configurer.py:212]:
```suggestion
        # TODO: Revisit this part when removing bg label -> it should be 1 because of 'background' label
        return (
            len(set(self.org_model_classes) & set(self.model_classes)) > 1
            and set(self.org_model_classes) != set(self.model_classes)
        )
```
How about to change code like this?

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py [hunk 7]
```diff
@@ -200,78 +196,20 @@ def configure_input_size(
...
-
-            if cfg.model.type in WEIGHT_MIX_CLASSIFIER:
-                cfg.model.task_adapt = ConfigDict(
-                    src_classes=self.org_model_classes,
-                    dst_classes=self.model_classes,
-                )
-            else:
-                logger.warning(f"Weight mixing for {cfg.model.type} is not yet supported!")
-
-            train_data_cfg.classes = self.model_classes
```
Comment (eunwoosh) [inline src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: It seems that these lines of code disappear. Is it ok?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your effort. Almost LGTM, but I have some comments. Please take a look.

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py (hunk 7 above)
Comment (eunwoosh) [inline src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: It seems that these lines of code disappear. Is it ok?

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py (hunk 2 above)
Comment (jaegukhyun) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py]: They are unused codes, so I deleted them

File: src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py (hunk 3 above)
Comment (jaegukhyun) [reply src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py]: I changed it to default_task_adapt to make easy to search. If you have any opinion for name, please notice me.

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py (hunk 1 above)
Comment (jaegukhyun) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py]: Good idea, I changed them

File: src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py (hunk 5 above)
Comment (jaegukhyun) [reply src/otx/algorithms/common/adapters/mmcv/clsincr_mixin.py]: I changed with your way, but I have to modify a little due to some pre-commit issues

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py (hunk 7 above)
Comment (jaegukhyun) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: They just print warning without any changes, so I just deleted them for simpler codes.

**Review by eunwoosh — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py [hunk 8]
```diff
@@ -284,49 +222,25 @@ def configure_loss(self, cfg):
...
-    def configure_hook(self, cfg):
-        """Update cfg.custom_hooks."""
-        super().configure_hook(cfg)
-        # Set unlabeled data hook
-        if self.training:
-            if cfg.data.get("unlabeled", False) and cfg.data.unlabeled.get("otx_dataset", False):
-                self.configure_unlabeled_dataloader(cfg)
+        Currently, only multi class classification supports adaptive repeat.
+        """
+        return self._is_multiclass(cfg)
```
Comment (harimkang) [inline src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: Will use_adaptive_repeat always be true if it's a multi class?
What if the user has a multi-class task and doesn't want to use adaptive_repeat?

File: src/otx/recipes/stages/classification/incremental.yaml [hunk 9]
```diff
@@ -29,7 +29,7 @@ lr_config:
   warmup_ratio: 0.333
 
 task_adapt:
-  type: "mpa"
+  type: "default_task_adapt"
```
Comment (harimkang) [inline src/otx/recipes/stages/classification/incremental.yaml:32]: Wouldn't something like `enable: True` be better in this case, rather than using type separately?
What do you think?
It doesn't seem to fit with type naming.

File: tests/unit/algorithms/segmentation/adapters/mmseg/test_mmseg_configurer.py [hunk 10]
```diff
@@ -280,40 +272,23 @@ def setup(self) -> None:
...
-    @e2e_pytest_unit
-    def test_configure_unlabeled_dataloader(self, mocker):
-        mocker_build_dataset = mocker.patch("otx.algorithms.segmentation.adapters.mmseg.configurer.build_dataset")
-        mocker_build_dataloader = mocker.patch("otx.algorithms.segmentation.adapters.mmseg.configurer.build_dataloader")
-        self.configurer.configure_hook(self.cfg)
-
-        mocker_build_dataset.assert_called_once()
-        mocker_build_dataloader.assert_called_once()
-        assert "ComposedDataLoadersHook" in [hook["type"] for hook in self.cfg.custom_hooks]
+        data_cfg = MPAConfig(
```
Comment (harimkang) [inline tests/unit/algorithms/segmentation/adapters/mmseg/test_mmseg_configurer.py:281]: What about renamed MPAConfig?
I don't think it's necessary to do this in this PR, but I thought I'd mention that MPA shouldn't be visible in OTX anymore.

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py (hunk 8 above)
Comment (sungmanc) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: Same with the `AdaptiveTrainingSchedule`, it will be automatically triggered. If users don't want to use, they could do by modifying recipe at the current status to remove the `CustomHook` named `AdaptiveRepeatDataHook`.

Comment (jaegukhyun) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: Current only multi class classification has two methods for adaptive training. First is adaptive training with adaptive epoch validation, which is common for the other tasks. The other is adaptive training with repeat sampler. This method enlarges iterations per epoch. Currently we all agreed that default operation of adaptive training for multi class classification is repeat sampler. Therefore, if user don't want this, then user should turn off adaptive_training flag.

Comment (harimkang) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: But shouldn't this function always come out as True, even if the user changes adaptive repeat to disable?

File: tests/unit/algorithms/segmentation/adapters/mmseg/test_mmseg_configurer.py (hunk 10 above)
Comment (jaegukhyun) [reply tests/unit/algorithms/segmentation/adapters/mmseg/test_mmseg_configurer.py:281]: Yes there are some mpa legacy in config and logger. I think they are beyond this PR. Let's handle this in another PR, but I'm not sure when it will be ready.

File: src/otx/algorithms/classification/adapters/mmcls/configurer.py (hunk 8 above)
Comment (harimkang) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: What I mean is that I don't think it's semantically correct for the `use_adaptive_repeat` function to `return is_multiclass` as it is. I think this function needs to be renamed, what do you think?

Comment (jaegukhyun) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: > But shouldn't this function always come out as True, even if the user changes adaptive repeat to disable?

Yes, but it does not effect training when users disable adaptive training

Comment (harimkang) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: > > But shouldn't this function always come out as True, even if the user changes adaptive repeat to disable?
> 
> Yes, but it does not effect training when users disable adaptive training

Yeah I was worried that from an API perspective, this is a public function and it would be confusing for users to see it always be true even when disabled.

Comment (jaegukhyun) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: > What I mean is that I don't think it's semantically correct for the use_adaptive_repeat function to return is_multiclass as it is.

I understand your concern, but it is refactoring PR, and this PR didn't change any logic for this. Therefore we need to discuss this and patch in anather PR

Previous code is
        task_adapt_hook = ConfigDict(
            type="TaskAdaptHook",
            src_classes=self.org_model_classes,
            dst_classes=self.model_classes,
            model_type=cfg.model.type,
            sampler_flag=sampler_flag,
            sampler_type=sampler_type,
            efficient_mode=efficient_mode,
            use_adaptive_repeat=is_multiclass,
            priority="NORMAL",
        )

Comment (sungmanc) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: I agree with @jaegukhyun and @harimkang . Current implementation is still temporary since we need to more implementation for multi-label and h-label and other tasks (det., seg.). So, it'll be refactored at my later PRs.

Comment (harimkang) [reply src/otx/algorithms/classification/adapters/mmcls/configurer.py:238]: Yes thanks for the clarification. I guess it's a function that will be removed in the near future.

**Comment by harimkang** [comment]
Thank you for your hard work. I have a few comments, but they are not critical. :)

File: src/otx/recipes/stages/classification/incremental.yaml (hunk 9 above)
Comment (jaegukhyun) [reply src/otx/recipes/stages/classification/incremental.yaml:32]: I'm not sure, since I don't have any background that why previous task adapt type is mpa. I just guess there could be another option for task_adapt. If there's not, maybe we can change type to enable. But it requires lots of changes, I want to make another PR for this and other things we discuss

Comment (harimkang) [reply src/otx/recipes/stages/classification/incremental.yaml:32]: yeah, no problem

**Review by goodsong81 — DISMISSED** [review]

**Review by harimkang — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]

File: src/otx/algorithms/common/adapters/mmcv/semisl_mixin.py (hunk 4 above)
Comment (jaegukhyun) [reply src/otx/algorithms/common/adapters/mmcv/semisl_mixin.py:25]: I revert this change since it raises errors for segmentation semisl, and hangs for detection semisl.

**Review by goodsong81 — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]
